src/playback-watcher.js

```diff
@@ -107,18 +131,81 @@ export default class PlaybackWatcher {
 
     const currentTime = this.tech_.currentTime();
     const buffered = this.tech_.buffered();
+    const atEndBuffer = !buffered.length ||
+      currentTime + Ranges.SAFE_TIME_DELTA >= buffered.end(buffered.length - 1);
+    // If current time is at the end of the final buffered region, then any playback
+    // stall is most likely caused by buffering in a low bandwidth environment. The tech
+    // should fire a `waiting` event in this scenario, but due to browser and tech
+    // inconsistencies. Calling `techWaiting_` here allows us to simulate
+    // responding to a native `waiting` event when the tech fails to emit one.
+
+    if (this.lastRecordedTime === currentTime && atEndBuffer) {
+      if (this.techWaiting_()) {
+        return;
+      }
+      const mpc = this.vhs.masterPlaylistController_;
+      const previousBytes = this.stallStartBytes_;
+
+      if (!previousBytes) {
+        this.tech_.off('timeupdate', this.checkForStallFix_);
+        this.tech_.on('timeupdate', this.checkForStallFix_);
+      }
+
+      this.stallStartBytes_ = mpc.mediaBytesTransferred_();
+
+      if (previousBytes && previousBytes !== this.stallStartBytes_) {
+        this.stallByteChanges_++;
+      }
+
+      // if we have yet to be stalled for 3 segment downloads
+      // or mpc has an error, do nothing
+      if (this.stallByteChanges_ < 3 || mpc.error) {
```

Comment:
Since `mediaBytesTransferred` is a summed loader stat between audio and video loaders, I believe that this will detect infinite buffering in the event that one loader has downloaded 3 segments before the other has downloaded 1, which may be a likely case for small audio segments and large video segments.

Reply:
I wonder if we should use the specific segment loaders `mediaBytesTransferred` then we know weather to try blacklisting audio or video 🤔